types/wlr_scene.c

```diff
@@ -426,10 +608,114 @@ void wlr_scene_render_output(struct wlr_scene *scene, struct wlr_output *output,
 			.output = output,
 			.damage = damage,
 		};
-		scene_node_for_each_node(&scene->node, lx, ly,
+		scene_node_for_each_node(&scene->node, -lx, -ly,
 			render_node_iterator, &data);
 		wlr_renderer_scissor(renderer, NULL);
 	}
 
 	pixman_region32_fini(&full_region);
 }
+
+static void scene_output_handle_destroy(struct wlr_addon *addon) {
+	struct wlr_scene_output *scene_output =
+		wl_container_of(addon, scene_output, addon);
+	wlr_scene_output_destroy(scene_output);
+}
+
+static const struct wlr_addon_interface output_addon_impl = {
+	.name = "wlr_scene_output",
+	.destroy = scene_output_handle_destroy,
+};
+
+struct wlr_scene_output *wlr_scene_output_create(struct wlr_scene *scene,
+		struct wlr_output *output) {
+	struct wlr_scene_output *scene_output = calloc(1, sizeof(*output));
+	if (scene_output == NULL) {
+		return NULL;
+	}
+
+	scene_output->damage = wlr_output_damage_create(output);
+	if (scene_output->damage == NULL) {
+		free(scene_output);
+		return NULL;
+	}
+
+	scene_output->output = output;
+	scene_output->scene = scene;
+	wlr_addon_init(&scene_output->addon, &output->addons, scene, &output_addon_impl);
+	wl_list_insert(&scene->outputs, &scene_output->link);
+
+	wlr_output_damage_add_whole(scene_output->damage);
+
+	return scene_output;
+}
+
+void wlr_scene_output_destroy(struct wlr_scene_output *scene_output) {
+	wlr_addon_finish(&scene_output->addon);
+	wl_list_remove(&scene_output->link);
+	free(scene_output);
+}
+
+void wlr_scene_output_set_position(struct wlr_scene_output *scene_output,
+		int lx, int ly) {
+	if (scene_output->x == lx && scene_output->y == ly) {
+		return;
+	}
+
+	scene_output->x = lx;
+	scene_output->y = ly;
+	wlr_output_damage_add_whole(scene_output->damage);
+}
+
+bool wlr_scene_output_commit(struct wlr_scene_output *scene_output) {
+	struct wlr_output *output = scene_output->output;
+
+	struct wlr_renderer *renderer = wlr_backend_get_renderer(output->backend);
+	assert(renderer != NULL);
+
+	bool needs_frame;
+	pixman_region32_t damage;
+	pixman_region32_init(&damage);
+	if (!wlr_output_damage_attach_render(scene_output->damage,
+			&needs_frame, &damage)) {
+		pixman_region32_fini(&damage);
+		return false;
+	}
+
+	if (!needs_frame) {
+		pixman_region32_fini(&damage);
+		wlr_output_rollback(output);
+		return true;
+	}
+
+	wlr_renderer_begin(renderer, output->width, output->height);
+
+	int nrects;
+	pixman_box32_t *rects = pixman_region32_rectangles(&damage, &nrects);
+	for (int i = 0; i < nrects; ++i) {
+		scissor_output(output, &rects[i]);
+		wlr_renderer_clear(renderer, (float[4]){ 0.0, 0.0, 0.0, 1.0 });
```

Comment:
Shouldnt the clear color be user defined?

Reply:
Eh, I don't see the need for it. If the caller really wants to set background color for whatever reason, they could create a solid color rectangle node, but again, I don't see the need for it.

Reply:
Yeah, to customize it users should just create a rect node with the color they want.